vtpm/src/char.rs

```diff
@@ -0,0 +1,267 @@
+        }
+        debug!("synchronous read from vTPM Socket");
+        let size = recv(self.ctrl_fd, buf, MsgFlags::empty()).map_err(|e| {
+            TPMCharError::TPMCharFailedRead(anyhow!(
+                "Failed to read from vTPM socket. Error Code {:?}",
+                e
+            ))
+        })?;
+        debug!("sync read completed");
+        Ok(size as usize)
+    }
+
+    pub fn send_full(&self, buf: &mut [u8], _len: usize) -> Result<usize> {
+        let iov = &[IoVec::from_slice(buf)];
+        let write_fd = self.write_msgfd;
+        let write_vec = &[write_fd];
+        let cmsgs = &[ControlMessage::ScmRights(write_vec)];
+        debug!("send full message");
+
+        // Send Ancillary data, along with cmds and data
+        let size = sendmsg(self.ctrl_fd, iov, cmsgs, MsgFlags::empty(), None).map_err(|e| {
+            TPMCharError::TPMCharFailedWrite(anyhow!(
+                "Failed to write to vTPM Socket. Error Code {:?}",
+                e
+            ))
+        })?;
+        Ok(size as usize)
+    }
+
+    pub fn chr_write(&mut self, buf: &mut [u8], len:usize) -> Result<usize> {
+        debug!("chr_write initialized");
+
+        if let Some(ref mut _sock) = self.stream {
+               let res = match self.state {
+                    ChardevState::ChardevStateConnected => {
+                        warn!("State Connected");
+                        let ret = self.send_full(buf, len)?;
+                        // swtpm will receive data Fd after a successful send
+                        // reset write_msgfd after a successful send
+                        self.write_msgfd = 0;
+                        ret
+                    },
+                    _ => return Err(TPMCharError::TPMCharFailedWrite(anyhow!(
+                        "TPM Socket was not in Connected State"))),
+            };
+            debug!("chr_write succeeded");
+
+            Ok(res)
+        } else {
+            return Err(TPMCharError::TPMCharFailedWrite(anyhow!("Stream for TPM Socket was not initialized")))
+        }
+    }
+
+    pub fn chr_read(&mut self, buf: &mut [u8], _len: usize) -> Result<usize> {
+        //Grab all response bytes so none is left behind
+        debug!("chr_read initialized");
+
+        let mut newbuf: &mut [u8] = &mut [0; TPM_TIS_BUFFER_MAX];
+
+        if let Some(ref mut sock) = self.stream {
+            let size:usize = sock.read(&mut newbuf).map_err(|e| TPMCharError::TPMCharFailedRead(anyhow!(
+                "Failed to read from vTPM Socket. Error Code {:?}",
+                e
+            )))?;
+            byte_copy(&newbuf, buf);
+            Ok(size)
+        } else {
+            return Err(TPMCharError::TPMCharFailedRead(anyhow!("Stream for TPM Socket was not initialized")))
+        }
+    }
+}
+/// This is the backend seen by frontend
+/// Actual Backend is SocketCharDev
+pub struct CharBackend {
+    pub chr: Option<SocketCharDev>,
+    fe_open: bool,
+}
+
+impl CharBackend {
```

Comment:
There is nothing tied to `Character`, is there?

Reply:
will address this as detailed in previous comment.